src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/BackupsApiController.java

```diff
@@ -25,34 +25,77 @@
+import org.niis.xroad.restapi.openapi.model.Backup;
+import org.niis.xroad.restapi.service.BackupFileNotFoundException;
+import org.niis.xroad.restapi.service.BackupsService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.security.access.prepost.PreAuthorize;
 import org.springframework.stereotype.Controller;
 import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.context.request.NativeWebRequest;
 
-import java.util.Optional;
+import java.util.List;
 
 /**
- * backups controller
+ * Backups controller
  */
 @Controller
 @RequestMapping("/api")
 @Slf4j
 @PreAuthorize("denyAll")
-@SuppressWarnings("checkstyle:HideUtilityClassConstructor")
 public class BackupsApiController implements BackupsApi {
 
-    public static final int MAX_FIFTY_ITEMS = 50;
+    public static final String GENERATE_BACKUP_INTERRUPTED = "generate_backup_interrupted";
 
-    private final NativeWebRequest request;
+    private final BackupsService backupsService;
+    @Autowired
+    public BackupsApiController(BackupsService backupsService) {
+        this.backupsService = backupsService;
+    }
+
+    @Override
+    @PreAuthorize("hasAuthority('BACKUP_CONFIGURATION')")
+    public ResponseEntity<List<Backup>> getBackups() {
+        List<Backup> backups = backupsService.getBackupFiles();
+
+        return new ResponseEntity<>(backups, HttpStatus.OK);
+    }
 
-    @org.springframework.beans.factory.annotation.Autowired
-    public BackupsApiController(NativeWebRequest request) {
-        this.request = request;
+    @Override
+    @PreAuthorize("hasAuthority('BACKUP_CONFIGURATION')")
+    public ResponseEntity<Void> deleteBackup(String filename) {
+        try {
+            backupsService.deleteBackup(filename);
+        } catch (BackupFileNotFoundException e) {
+            throw new ResourceNotFoundException(e);
+        }
+
+        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+    }
+
+    @Override
+    @PreAuthorize("hasAuthority('BACKUP_CONFIGURATION')")
+    public ResponseEntity<Resource> downloadBackup(String filename) {
+        byte[] backupFile = null;
+        try {
+            backupFile = backupsService.readBackupFile(filename);
+        } catch (BackupFileNotFoundException e) {
+            throw new ResourceNotFoundException(e);
+        }
+        return ApiUtil.createAttachmentResourceResponse(backupFile, filename);
     }
 
     @Override
-    public Optional<NativeWebRequest> getRequest() {
-        return Optional.ofNullable(request);
+    @PreAuthorize("hasAuthority('BACKUP_CONFIGURATION')")
+    public ResponseEntity<Backup> addBackup() {
+        try {
+            Backup backup = backupsService.generateBackup();
+            return new ResponseEntity<>(backup, HttpStatus.ACCEPTED);
```

Comment:
Return code is changed to `CREATED`. However, I did not use ApiUtil-added details since we don't have an endpoint to access the metadata of a specific backup. This is aligned with other similar endpoints, e.g. [adding a configured timestamping service](https://github.com/nordic-institute/X-Road/blob/0626f471cf11671ce7295a782600039509c45719/src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/SystemApiController.java#L132) (`POST /api/timestamping-services`).